iOS,macOS: Migrate logging to Logger/FlutterLogger

cbracken opened the pull request:
In new Swift code in the iOS/macOS embedder, we currently lack core features currently provided by shared internal C++ library code in FML. This provides a logging implementation that can be used across the iOS/macOS embedders, whether in Swift, Objective-C, or C++ code, and migrates the iOS and macOS embedders to use it.

This patch also introduces gn support for authoring [Swift Testing][] tests.
* Introduces a `SwiftTestingRunner` class that launches the Swift Testing test runner, which runs all tests annotated with `@Test` linked into the current executable.
* Introduces a `SwiftTestingRunnerMain` class that implements a main entrypoint that launches the `SwiftTestingRunner`.
* Introduces a `gn` `source_set` target (`//flutter/shell/platform/darwin/common:swift_testing_main`) that includes the Swift Testing main entrypoint, such that `executable` targets can add it to their deps to automatically be built as a Swift Testing runner.
* Adds a `framework_common_swift_unittests` executable target for the common framework unit tests, and wires that into the build and `run_tests.py` for macOS.

Issue: https://github.com/flutter/flutter/issues/168564
Issue: https://github.com/flutter/flutter/issues/144791

[Swift Testing]: https://developer.apple.com/xcode/swift-testing/


## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

cbracken committed:
iOS,macOS: Add Logger/FlutterLogger

In new Swift code in the iOS/macOS embedder, we currently lack core
features currently provided by shared internal C++ library code in FML.
This provides a logging implementation that can be used across the
iOS/macOS embedders, whether in Swift, Objective-C, or C++ code.

Issue: https://github.com/flutter/flutter/issues/168564

cbracken committed:
Migrate from FML_LOG to FlutterLogger

jmagman commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,116 @@
+
+final class SyslogOutputWriter: OutputWriter {
+  func writeLine(_ message: String) {
+    // TODO(cbracken): replace this with os_log-based approach.
```
I don't know how feasible this is, but I would love to replace this with proper string formatting, since static strings aren't redacted in logs, but dynamic ones are:
https://developer.apple.com/documentation/os/generating-log-messages-from-your-code?#Redact-Sensitive-User-Data-from-a-Log-Message

cbracken replied:
Agreed -- this one tries to purposely keep things super stupid and port line for line what we do today, but once we have this all centralised here, then we absolutely should look at trying for os_log one more time and getting all the nice stuff like that!

I originally tried to export a varargs version of this but `@objc` bridging doesn't support it, sadly. We could make a Swift-only API alongside these though so at least Swift code would have it!

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
````diff
@@ -0,0 +1,116 @@
+/// ```
+@objc(FlutterLogger) public final class Logger: NSObject {
+  /// The shared singleton instance of the logger.
+  @objc public static let instance = Logger()
````
nit: `shared`

hellohuanlin replied:
also if you wanna make it unit testable, you will have to make it a `var`: 

```
@objc public internal(set) static var instance = Logger()
```

Then do a `@testable` import in unit test.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,116 @@
+  @objc public static let instance = Logger()
+
+  /// Logs a message at `LogLevel.info`.
+  @objc public static func logInfo(_ message: String) {
```
nit: we can have these `func logFOO(message:)` under an extension

cbracken replied:
I like it! Will do.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  /// Logs a message unconditionally.
+  @objc public static func logDirect(_ message: String) {
```
do we actually want this? maybe give an example in the comment?

cbracken replied:
It's used in this patch.
https://github.com/flutter/flutter/pull/168568/files#diff-1e9b73bf9e1fc12da0ff2e6ec30116563ecbd5996010ef9c7166155439f0da20R78

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  var logLevel = LogLevel.info
+  let outputWriter: OutputWriter
```
can these 2 be private?

cbracken replied:
Yep! Will do.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,116 @@
+#endif
+  }
+
+  private func log(level: LogLevel, _ message: String) {
```
nit: `log(_ message, at level: LogLevel)`

cbracken replied:
I wrote it this way to mirror Apple's `os_log` function:
https://developer.apple.com/documentation/os/3580306-os_log#

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,116 @@
+  func writeLine(_ message: String) {
+    fputs(message, stdout)
+    fputs("\n", stdout)
+    fflush(stdout)
```
can we just do `print`? https://developer.apple.com/documentation/swift/print(_:separator:terminator:)

cbracken replied:
Print has platform-specific behaviour and is not documented to write directly to `stdout`. You can think of this as implementing a peer to `print`.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,116 @@
+  let outputWriter: OutputWriter
+
+  private override init() {
+#if os(iOS)
```
maybe comment on why using syslog on ios and stdout on macos?

cbracken replied:
Done.

cbracken commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
````diff
@@ -0,0 +1,116 @@
+/// ```
+@objc(FlutterLogger) public final class Logger: NSObject {
+  /// The shared singleton instance of the logger.
+  @objc public static let instance = Logger()
````
This can be private. I used to not have the `setLogLevel` method, but now that we have it, it's unnecessary.

cbracken committed:
Allow Swift DATA_CONST,DATA symbol table entries

Previously updated for non-Swift in flutter/flutter#167737.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
````diff
@@ -0,0 +1,125 @@
+/// Logger.logLevel = .warning  // Only show warnings and above
+/// Logger.logError("Failed to load asset: \(assetKey)")
+/// ```
+@objc(FlutterLogger) public final class Logger: NSObject {
````
FYI another design I mentioned yesterday was something like this (semi-pseudo code)
```
class Logger {
  static var shared: LoggerImpl = DefaultLoggerImpl()
  static func setLogLevel(level: LogLevel) {
    shared.logLevel = level
  }
  static func logInfo(msg: String) {
    shared.log(level: .info, msg: msg)
  }
  static func logWarning(msg: String) {
    shared.log(level: .warning, msg: msg)
  }
  // ... 
}
protocol LoggerImpl {
  var logLevel { get set }
  func log(level: LogLevel, _ message: String)
}
class DefaultLoggerImpl {
  var logLevel: LogLevel
  func log(level: LogLevel, _ message: String) {
    // actual logging logic here (syslog or stdio)
  }
}

```
Then in unit test, you can swap out the whole `LoggerImpl` to a `MockLoggerImpl`. 
```
class MockLoggerImpl: LoggerImpl {
  var logLevel: LogLevel
  var logStub: ((LogLevel, String) -> Void)?
  func log(level: LogLevel, _ message: String) {
    logStub?(level, message)
  }
}
```
Then to use this mock: 
```
let mock = MockLoggerImpl()
mock.logStub = { _, _ in 
  fooEventShouldBeLoggedExpectation.fulfill()
}
Logger.shared = mock
doSomethingThatWillTriggerThisLog()
```

cbracken replied:
I think what you've outlined is exactly what this patch does, except that your protocol is named `LoggerImpl` and mine is named `OutputWriter`, and yours uses a callback whereas mine uses a method implementation.

hellohuanlin replied:
> I think what you've outlined is exactly what this patch does

Almost - see the level part is also part of the LoggerImpl protocol. 

> except that your protocol is named `LoggerImpl` and mine is named `OutputWriter`, and yours uses a callback whereas mine uses a method implementation.

The `LoggerImpl` protocol doesn't use a callback (or am I missing something here?)

cbracken replied:
If you move the `logLevel` to the `LoggerImpl` then you're not testing the logic if you write a test `LoggerImpl`, and there really is nothing left to test at that point. I think I'll stick with the current design.

hellohuanlin replied:
> If you move the `logLevel` to the `LoggerImpl` then you're not testing the logic if you write a test LoggerImpl, and there really is nothing left to test at that point. 

Are you referring to (1) testing the logger itself? or referring to (2) testing the code that uses on the logger? 

If you are referring to (1), we can focus on testing `DefaultLoggerImpl`. The `Logger` class itself is just a static wrapper around `LoggerImpl`, so testing it would be trivial. 

If you are referring to (2), we can mock both `log(level:msg:)` method and `logLevel` setter. 

> I think I'll stick with the current design.

To clarify, I'm not against the current design - I just wanted to provide an alternative and to exchange ideas. Maybe i'm missing something here - I'm still confused about the `callback` vs `method implementation` part in your previous comment.

cbracken replied:
> Are you referring to (1) testing the logger itself? or referring to (2) testing the code that uses on the logger?

The current design is better at both:
(1) In this case, there's literally no point in writing a test if all logging functionality is hardcoded in the logger (both the prioritisation and the output destination, such as stdout), because now, for no reason at all, we have to make life really hard on ourselves and attempt to hijack that destination, which almost certainly means adding an injectable OutputWriter equivalent, in which case this solution already has that.
(2) In this case, which is something we almost never do in current codebase, since embedder logging is purely debug/informational and not something that forms an interface contract, injecting an OutputWriter that writes to string solves the problem with way less code than the alternative being proposed.

hellohuanlin replied:
Thanks for the explanation. I think I get what you mean, but just double-check that I *actually* understand: 

Implementations inside `SyslogOutputWriter` and `StdioOutputWriter` classes cannot be easily unit tested (have to hijack `stdout`), so we won't test them. However, `logLevel` part of the logic can is easily testable, so we want to separate these 2 parts. Did I understand correctly?

cbracken replied:
Yep correct. We don't really need to test either of those since both of these are just a wrapper around `syslog` and `fputs/fflush`, which have been around for almost half a century at this point, so we can be pretty sure they work.

hellohuanlin replied:
Make sense! I somehow missed the part that we don't test those. This is resolved. I will take another pass for this PR.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,132 @@
+/// They are ordered by increasing severity.
+@objc(FlutterLogLevel) public enum LogLevel: Int {
+  /// Informational messages that are helpful for tracing application flow.
+  case info = 0
```
nit: i believe `= 0` is the default, so can omit here.

cbracken replied:
It's default, but because these are used in integer comparisons, it's potentially useful for readability to write them out explicitly, as we also do in [log_level.h](https://github.com/flutter/flutter/blob/master/engine/src/flutter/fml/log_level.h), even though enums in C/C++ (and most languages) are 0-based.

While the goal is to replace FML logging, realistically a bunch of people work across embedders in Flutter and it's likely useful to keep our log levels and their meanings in sync, and if we're going to do that, then may as well keep the values matching. 

Might be useful though to link back to fml/log_level.h in a comment though.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,132 @@
+  public let logLevel = LogLevel.info
+
+  public init(outputWriter: OutputWriter, logLevel: LogLevel) {
+    self.outputWriter = outputWriter
```
Looks like this is mistakenly checked-in? (`logLevel` parameter is not used). I will pause the review and look again when it's ready.

cbracken replied:
Not checked in but good catch -- missed that line before pushing.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,133 @@
+  /// Sets the minimum log level.
+  @objc public static var outputWriter: OutputWriter {
+    get { return shared.outputWriter }
+    set(outputWriter) { shared = Logger(outputWriter: outputWriter, logLevel: shared.logLevel) }
```
is there any reason we make `shared` mutable and Logger's properties immutable?

cbracken replied:
The alternative is tomato all of Logger's properties mutable and `shared` immutable. These are equivalent designs, but less mutability is better.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,133 @@
+  /// Sets the minimum log level.
+  @objc public static var logLevel: LogLevel {
+    get { return shared.logLevel }
+    set(logLevel) { shared = Logger(outputWriter: shared.outputWriter, logLevel: logLevel) }
```
nit: can use `newValue` in setter

cbracken replied:
Done.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,133 @@
+
+  /// Logs a message unconditionally.
+  @objc public static func logDirect(_ message: String) {
+    shared.outputWriter.writeLine(level: .important, message)
```
maybe document why use `.important`? (i assume it's because it's the highest level besides `fatal`).

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift:
```diff
@@ -0,0 +1,133 @@
+final class StdoutOutputWriter: OutputWriter {
+  func writeLine(level: LogLevel, _ message: String) {
+    fputs(message, stdout)
+    fputs("\n", stdout)
```
does the new line already flush it?

cbracken replied:
No.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/common/framework/Source/LoggerTestUtils.swift:
```diff
@@ -0,0 +1,26 @@
+  public var lastLevel: LogLevel!
+  @objc public var lastLine: String!
+
+  public func writeLine(level: LogLevel, _ message: String) {
```
should we expose this func to objc as well?

cbracken replied:
There's no reason to expose this to Obj-C since no test should ever call it; this function is purely for the logger, which is written in Swift.

cbracken committed:
Liceeeeeeeeeeeeeeeences

cbracken committed:
Migrate to public test runner API

This migrates off the private API for Swift Package Manager, and onto
the official v0 public ABI.
